File: #0005.longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {

        int size = s.size();
        string result = "";
        vector<vector<int>> v(2, vector(size+1, 0));
        string s2(s.rbegin(), s.rend());
        for (int i = 1; i <= size; i++) {
            for (int j = 1; j <= size; j++) {

                if (s[i-1] == s2[j-1]) {
                    v[i%2][j] = v[(i+1)%2][j-1] + 1;
                } else {
                    v[i%2][j] = 0;
                }
                
                if (v[i % 2][j] > result.size() && i + j == v[i % 2][j] + size) {
                    result = s.substr(i-v[i % 2][j], v[i % 2][j]);
                }
            }
        }
        
        return result;
    }
};

File: #0005.longest-palindromic-substring.js
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function(s) {
    let result = s[0] || "";
    const search = (left, right) => {
        while (left < s.length
            && left - right.length - 1 > -1
            && s[left] === s[left - right.length - 1]) {
            right = s[left] + right + s[left];
            left++;
        }

        result = result.length < right.length ? right : result;
    }

    for (let i = 1; i < s.length; i++) {
        if (s[i - 1] === s[i]) {
            search(i + 1, s[i - 1] + s[i]);
        }
        search(i + 1, s[i]);
    }
    return result;
};
